### PR TITLE
Publishing automation - start

### PR DESCRIPTION
**Pull request by Bnaya**
* Code is still plain js, but fully covered by typescript via jsdocs
* Refactor code to be more programatic
* Added utils function that can check if we need to build & publish new
version
* Added dummy circleci file
* Some vscode configs

implements : https://github.com/hot-loader/react-dom/issues/9

### CODE REVIEW

**Comment by Bnaya** [comment]
@theKashey I would like your preliminary opinion regarding my work
code styles, infra decisions, etc 
The most interesting part is
`auto-publish/utils.js `

I've also sent you a request from circleci to give me access to add this project there,
And then i will start playing with the periodic stuff (That will basically run the code we have here)

File: auto-publish/utils.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+  const originalDir = path.resolve(wd, "source");
+  const targetDir = path.resolve(wd, "target");
+
+  await fs.mkdirp(originalDir);
+  await fs.mkdirp(targetDir);
+
+  // create new local package.json for isolation
+  await execFile("yarn", ["init", "-y"], originalDir);
+  // install react so avoid yarn peer warnings
+  await execFile(
```
Comment (theKashey) [inline auto-publish/utils.js]: why not install `react` and `react-dom` in one run?

File: auto-publish/utils.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
+const child_process = require("child_process");
+const semver = require("semver");
+const path = require("path");
+const fs = require("fs-extra");
+const OUR_RC_DIST_TAG = "rc";
+const basePackageJSON = require("../source/package.base.json");
+const {
+  default: { patch }
+  // @ts-ignore
+} = require("react-hot-loader/webpack");
```
Comment (theKashey) [inline auto-publish/utils.js]: `react-hot-loader` then also have to be auto-updated.

File: jest.config.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
+// For a detailed explanation regarding each configuration property, visit:
```
Comment (theKashey) [inline jest.config.js]: Let's keep it simple

**Comment by theKashey** [comment]
I'll port [RHL+react-dom integration test](https://github.com/gaearon/react-hot-loader/blob/master/test/hot/react-dom.integration.spec.js) to verify the result.
Even if RHL is testing `patch+react-dom` - it does in it a separate loop, and integration might break with `react-dom` update.

File: auto-publish/utils.js (hunk 2 above)
Comment (Bnaya) [reply auto-publish/utils.js]: We may track the version change of react-hot-loader & react-dom, and make new version when each of them changes,
But which version number should we give our artifact? we won't be able to strictly follow react-dom no more

File: jest.config.js (hunk 3 above)
Comment (Bnaya) [reply jest.config.js]: this is the default generated config of jest :)
You mean you want to remove all of the comments?

Comment (theKashey) [reply jest.config.js]: Yeah ;)

File: auto-publish/utils.js (hunk 2 above)
Comment (theKashey) [reply auto-publish/utils.js]: 16.8.6-xxxx?

Comment (Bnaya) [reply auto-publish/utils.js]: you mean:
`react.dom.version-rhl.version` ? and then we can split  on `-` and see the version in rhl or dom have changed?

It means we somewhat lose semver semantics, but i can't think about any other good option to save state of last released versions

**Comment by Bnaya** [comment]
it has manual approval. super cool.
https://circleci.com/docs/2.0/triggers/#manual-approval

File: auto-publish/utils.js (hunk 2 above)
Comment (Bnaya) [reply auto-publish/utils.js]: Done, and please see how the new code handling the version

**Comment by Bnaya** [comment]
@theKashey friendly ping :)

**Comment by theKashey** [comment]
Ouch. Pinging on weekends not always work well :)

**Comment by Bnaya** [comment]
I'm SORRY!
I'm from Israel and i forgot its still your weekend :(

**Comment by Bnaya** [comment]
So what's next?

**Pull request merged**